src/jit/hwintrinsicxarch.cpp

```diff
@@ -445,16 +570,40 @@ GenTree* Compiler::impHWIntrinsic(NamedIntrinsic        intrinsic,
     {
         return impUnsupportedHWIntrinsic(CORINFO_HELP_THROW_PLATFORM_NOT_SUPPORTED, method, sig, mustExpand);
     }
-    else if (category == HW_Category_IMM)
+    // Avoid checking stacktop for 0-op intrinsics
+    if (sig->numArgs > 0 && isImmHWIntrinsic(intrinsic, impStackTop().val))
     {
         GenTree* lastOp = impStackTop().val;
-        if (!lastOp->IsCnsIntOrI() && !mustExpand)
+        // The imm-HWintrinsics that do not accept all imm8 values may throw exception or transform to other
+        // intrinsics when the imm argument is not in the valid range
+        if ((flags & HW_Flag_FullRangeIMM) == 0)
         {
-            // When the imm-argument is not a constant and we are not being forced to expand, we need to
-            // return nullptr so a GT_CALL to the intrinsic method is emitted instead. The
-            // intrinsic method is recursive and will be forced to expand, at which point
-            // we emit some less efficient fallback code.
-            return nullptr;
+            // Throw ArgumentOutOfRangeException if the imm-value is not in the valid range
+            if (!mustExpand && lastOp->IsCnsIntOrI() &&
+                lastOp->AsIntCon()->IconValue() > immUpperBoundOfHWIntrinsic(intrinsic))
+            {
+                return nullptr;
+            }
+
+            // TODO-XARCH-CQ repleace imm-intrinsic by certain lower-latency intrinsic based-on imm-values.
```

Comment:
nit: I think this is one of the optimizations we might not want to make since it violates the contract of emitting a particular instruction.

Reply:
A performance analyzer would be well-suited to calling this type of optimization out instead.

Reply:
C++ compilers (e.g., Clang, ICC) is making this optimization.

Reply:
Yes, they also make other optimizations such as using `UNPCKHPS` instead of `SHUFPS` in some cases.

However, I believe we had determined that we should keep this simple and emit exactly what the user requested under the assumption that they had manually tuned the code. (CC. @CarolEidt, in case I misunderstood).

The exception to this was the fallback case where we:
* Emit a jump table that will ultimately call the exact instruction
or 
* Resolve it to the same instruction, but with a different encoding which did not require the immediate parameter (some shift instructions)

Reply:
> user requested under the assumption that they had manually tuned the code. 

Actually, I still think we should enable some optimizations for HW intrinsics (e.g., CSE) in the future. HW intrinsic is not only used as "higher-level assembly" to optimize a few functions, users can write complex fully-vectorized programs.

Reply:
IMO this should be solved by compiler hints. Developers will know when they can allow optimizations to be on and where optimizations should be always off. However, I would defer implementing this after first intriniscs implementation is shipped.

Reply:
I suspect that our strategy here will evolve over time, but I think we should lean toward a faithful instantiation of the requested intrinsics. Over time we might want to consider what transformations or optimizations might be supported, but that makes sense as a future item.

Reply:
I would urge you to just remove this comment.